File: frontend/src/types/gameWallet.ts
/** Server-authoritative game loyalty wallet (R21). Credits are integer strings. */

export interface GameWallet {
  user_id: number
  account_balance: string
  credits: string
  free_spins_remaining: number
  loyalty_enabled: boolean
  loyalty_available: boolean
  checkin_credits: string
  slot_bet_credits: string
  daily_reward_limit: number
  checked_in_today: boolean
  today_checkin_at?: string | null
  local_date: string
  /** Legacy R20 fields; always false/zero in R21. */
  exchange_enabled?: boolean
  exchange_available?: boolean
  exchange_rate?: string
  daily_limit?: string
  daily_exchanged?: string
  daily_remaining?: string | null
  pending_slot_bonus?: GameWalletSlotBonusRound
  wallet_created_at: string
  wallet_updated_at: string
}

export interface GameWalletCheckinResult {
  id: string
  user_id: number
  local_date: string
  credits_awarded: string
  credits_before: string
  credits_after: string
  idempotency_key: string
  idempotent_replay?: boolean
  created_at: string
}

export interface GameCreditGiftRequest {
  recipient_email: string
  credits: string
}

export interface GameCreditGiftResult {
  id: string
  sender_user_id: number
  recipient_user_id: number
  recipient_email: string
  credits: string
  sender_credits_before: string
  sender_credits_after: string
  recipient_credits_before: string
  recipient_credits_after: string
  idempotency_key: string
  idempotent_replay?: boolean
  created_at: string
}

export interface GameWalletSpinRequest {
  bet_credits?: string
}

export type GameFruitDoor = 'bar' | '77' | 'star' | 'watermelon' | 'bell' | 'mango' | 'orange' | 'apple'

export type GameFruitBets = Record<GameFruitDoor, string>

export interface GameFruitRewardStop {
	index: number
	symbol: GameFruitDoor
	size: 'big' | 'small'
	multiplier: string
}

export interface GameFruitOutcome {
	kind: 'fruit' | 'send_light' | 'little_mary' | 'jackpot'
	symbol?: GameFruitDoor
	size?: 'big' | 'small'
	multiplier?: string
	extra_stops?: GameFruitRewardStop[]
	little_mary_stops?: GameFruitRewardStop[]
	jackpot_tier?: 'bronze' | 'silver' | 'gold'
	jackpot_multiplier?: string
}

export interface GameFruitSpinResult {
	id: string
	user_id: number
	bets: GameFruitBets
	total_bet: string
	stop_index: number
	outcome: GameFruitOutcome
	payout: string
	credits_before: string
	credits_after: string
	paytable_version: string
	idempotency_key: string
	idempotent_replay?: boolean
	created_at: string
}

export interface GameWalletSlotBonusRound {
  round_id: string
  token: string
  expires_at: string
  choices: number
  kind: 'free_spins'
  bet_credits: string
  bonus_count: number
  bonus_multiplier: number
  options: GameWalletSlotBonusOption[]
  status?: string
}

export interface GameWalletSlotBonusOption {
  choice: number
  spins: number
  multiplier: number
}

export interface GameWalletSlotBonusClaimRequest {
  token: string
  choice: number
}

export interface GameWalletSlotBonusClaimResult {
  round_id: string
  kind: 'free_spins'
  choice: number
  reward_credits: string
  credits_after: string
  claimed_at: string
  idempotent_replay?: boolean
  spins: number
  multiplier: number
  bonus_multiplier: number
  total_base_payout: string
  spin_results: GameWalletSlotBonusSpinResult[]
}

export interface GameWalletWinningLine {
  line_index: number
  symbol: string
  count: number
  multiplier: string
  payout: string
  /** [reel, row] pairs */
  positions: Array<[number, number]>
}

export interface GameWalletSpinResult {
  id: string
  user_id: number
  bet_credits: string
  payout_credits: string
  used_free_spin: boolean
  /** 3 rows x 5 reels of symbol codes */
  grid: string[][]
  stops: number[]
  winning_lines: GameWalletWinningLine[]
  scatter_count: number
  /** R23 BONUS count; absent on legacy R22 payloads. */
  bonus_count?: number
  free_spins_awarded: number
  free_spins_remaining: number
  credits_before: string
  credits_after: string
  paytable_version: string
  reel_strip_version: string
  idempotency_key: string
  idempotent_replay?: boolean
  created_at: string
  bonus_round?: GameWalletSlotBonusRound
}

export type GameWalletSlotBonusSpinResult = Partial<Omit<GameWalletSpinResult, 'grid' | 'winning_lines'>> & {
  grid: string[][]
  winning_lines?: Array<Omit<GameWalletWinningLine, 'multiplier' | 'payout'> & {
    multiplier: string | number
    payout: string | number
  }>
  total_payout?: string | number
}

export interface GameLeaderboardEntry {
  rank: number
  user_display: string
  score: string
  achieved_at: string
  is_current_user: boolean
}

export interface GameLeaderboardResult {
  game_id: string
  local_date: string
  items: GameLeaderboardEntry[]
  my_entry?: GameLeaderboardEntry | null
}

export interface GameLeaderboardScoreResult {
  game_id: string
  score: string
}

export interface GameLeaderboardQuery {
  game_id: string
  limit?: number
}

export interface GameWalletRewardItem {
  id: string
  title: string
  credit_cost: string
  voucher_value: string
  daily_stock: number
  expiry_days: number
  remaining_stock: number | null
  claimed_today: number
}

export interface GameWalletRewardList {
  items: GameWalletRewardItem[]
  daily_reward_limit: number
  claimed_today: number
  loyalty_enabled: boolean
  loyalty_available: boolean
}

export interface GameWalletRewardClaimRequest {
  reward_id: string
}

export interface GameWalletRewardClaimResult {
  id: string
  user_id: number
  reward_id: string
  title: string
  credit_cost: string
  voucher_value: string
  redeem_code: string
  claim_local_date: string
  credits_before: string
  credits_after: string
  expires_at?: string | null
  idempotency_key: string
  idempotent_replay?: boolean
  created_at: string
}

/** Ledger entry types from the server. */
export type GameWalletEntryType =
  | 'checkin'
  | 'slot_bet'
  | 'slot_payout'
  | 'slot_bonus'
	| 'fruit_bet'
	| 'fruit_payout'
	| 'gift_sent'
	| 'gift_received'
  | 'reward_claim'
  | string

export interface GameWalletTransaction {
  id: string
  user_id: number
  entry_type: GameWalletEntryType
  amount: string
  credits_before: string
  credits_after: string
  reference_type?: string
  reference_id?: string | null
  idempotency_key: string
  metadata?: unknown
  created_at: string
}

export interface GameWalletTransactionList {
  items: GameWalletTransaction[]
  next_before_id: string | null
}

export interface GameWalletTransactionQuery {
  limit?: number
  before_id?: string
}

/** Admin catalog row before serialization to exact JSON string. */
export interface GameLoyaltyRewardCatalogItem {
  id: string
  title: string
  credit_cost: string
  voucher_value: string
  daily_stock: number
  expiry_days: number
  enabled: boolean
  titles?: Record<string, string>
}

File: backend/internal/handler/dto/game_wallet.go
package dto

import (
	"bytes"
	"encoding/json"
	"fmt"
	"time"
)

// GameWalletNumber preserves the exact JSON token so both string and legacy
// numeric request values can be validated without passing through float64.
type GameWalletNumber string

func (n *GameWalletNumber) UnmarshalJSON(data []byte) error {
	raw := bytes.TrimSpace(data)
	if len(raw) == 0 || bytes.Equal(raw, []byte("null")) {
		return fmt.Errorf("amount must not be null")
	}
	if raw[0] == '"' {
		var value string
		if err := json.Unmarshal(raw, &value); err != nil {
			return err
		}
		*n = GameWalletNumber(value)
		return nil
	}
	*n = GameWalletNumber(string(raw))
	return nil
}

func (n *GameWalletNumber) StringPointer() *string {
	if n == nil {
		return nil
	}
	value := string(*n)
	return &value
}

// Deprecated R20 exchange request retained only so old clients get a clean 410
// from the removed route rather than a bind error storm.
type GameWalletExchangeRequest struct {
	Direction     string            `json:"direction"`
	BalanceAmount *GameWalletNumber `json:"balance_amount"`
	CreditsAmount *GameWalletNumber `json:"credits_amount"`
}

type GameWalletState struct {
	UserID             int64      `json:"user_id"`
	AccountBalance     string     `json:"account_balance"`
	Credits            string     `json:"credits"`
	FreeSpinsRemaining int        `json:"free_spins_remaining"`
	LoyaltyEnabled     bool       `json:"loyalty_enabled"`
	LoyaltyAvailable   bool       `json:"loyalty_available"`
	CheckinCredits     string     `json:"checkin_credits"`
	SlotBetCredits     string     `json:"slot_bet_credits"`
	SlotBetOptions     []string   `json:"slot_bet_options"`
	DailyRewardLimit   int        `json:"daily_reward_limit"`
	CheckedInToday     bool       `json:"checked_in_today"`
	TodayCheckinAt     *time.Time `json:"today_checkin_at,omitempty"`
	LocalDate          string     `json:"local_date"`
	// Legacy exchange fields kept for older UIs; always false/zero in R21.
	ExchangeEnabled   bool                  `json:"exchange_enabled"`
	ExchangeAvailable bool                  `json:"exchange_available"`
	ExchangeRate      string                `json:"exchange_rate"`
	DailyLimit        string                `json:"daily_limit"`
	DailyExchanged    string                `json:"daily_exchanged"`
	DailyRemaining    *string               `json:"daily_remaining"`
	WalletCreatedAt   time.Time             `json:"wallet_created_at"`
	WalletUpdatedAt   time.Time             `json:"wallet_updated_at"`
	PendingSlotBonus  *GameWalletBonusRound `json:"pending_slot_bonus,omitempty"`
}

type GameWalletCheckinResult struct {
	ID               string    `json:"id"`
	UserID           int64     `json:"user_id"`
	LocalDate        string    `json:"local_date"`
	CreditsAwarded   string    `json:"credits_awarded"`
	CreditsBefore    string    `json:"credits_before"`
	CreditsAfter     string    `json:"credits_after"`
	IdempotencyKey   string    `json:"idempotency_key"`
	IdempotentReplay bool      `json:"idempotent_replay,omitempty"`
	CreatedAt        time.Time `json:"created_at"`
}

type GameWalletSpinRequest struct {
	// BetCredits is optional. When set it must equal the configured slot bet.
	// Free spins are server-owned and cannot be requested by the client.
	BetCredits *GameWalletNumber `json:"bet_credits"`
}

type GameFruitSpinRequest struct {
	Bets map[string]GameWalletNumber `json:"bets" binding:"required"`
}

type GameFruitRewardStop struct {
	Index      int    `json:"index"`
	Symbol     string `json:"symbol"`
	Size       string `json:"size"`
	Multiplier string `json:"multiplier"`
}

type GameFruitOutcome struct {
	Kind              string                `json:"kind"`
	CenterIndex       int                   `json:"center_index,omitempty"`
	Symbol            string                `json:"symbol,omitempty"`
	Size              string                `json:"size,omitempty"`
	Multiplier        string                `json:"multiplier,omitempty"`
	ExtraStops        []GameFruitRewardStop `json:"extra_stops,omitempty"`
	LittleMaryStops   []GameFruitRewardStop `json:"little_mary_stops,omitempty"`
	JackpotTier       string                `json:"jackpot_tier,omitempty"`
	JackpotMultiplier string                `json:"jackpot_multiplier,omitempty"`
}

type GameFruitSpinResult struct {
	ID               string            `json:"id"`
	UserID           int64             `json:"user_id"`
	Bets             map[string]string `json:"bets"`
	TotalBet         string            `json:"total_bet"`
	StopIndex        int               `json:"stop_index"`
	Outcome          GameFruitOutcome  `json:"outcome"`
	Payout           string            `json:"payout"`
	CreditsBefore    string            `json:"credits_before"`
	CreditsAfter     string            `json:"credits_after"`
	PaytableVersion  string            `json:"paytable_version"`
	IdempotencyKey   string            `json:"idempotency_key"`
	IdempotentReplay bool              `json:"idempotent_replay,omitempty"`
	CreatedAt        time.Time         `json:"created_at"`
}

type GameWalletWinningLine struct {
	LineIndex  int      `json:"line_index"`
	Symbol     string   `json:"symbol"`
	Count      int      `json:"count"`
	Multiplier string   `json:"multiplier"`
	Payout     string   `json:"payout"`
	Positions  [][2]int `json:"positions"`
}

type GameWalletSpinResult struct {
	ID                 string                  `json:"id"`
	UserID             int64                   `json:"user_id"`
	BetCredits         string                  `json:"bet_credits"`
	PayoutCredits      string                  `json:"payout_credits"`
	UsedFreeSpin       bool                    `json:"used_free_spin"`
	Grid               [][]string              `json:"grid"`
	Stops              []int                   `json:"stops"`
	WinningLines       []GameWalletWinningLine `json:"winning_lines"`
	ScatterCount       int                     `json:"scatter_count"`
	BonusCount         int                     `json:"bonus_count"`
	FreeSpinsAwarded   int                     `json:"free_spins_awarded"`
	FreeSpinsRemaining int                     `json:"free_spins_remaining"`
	CreditsBefore      string                  `json:"credits_before"`
	CreditsAfter       string                  `json:"credits_after"`
	PaytableVersion    string                  `json:"paytable_version"`
	ReelStripVersion   string                  `json:"reel_strip_version"`
	IdempotencyKey     string                  `json:"idempotency_key"`
	IdempotentReplay   bool                    `json:"idempotent_replay,omitempty"`
	CreatedAt          time.Time               `json:"created_at"`
	BonusRound         *GameWalletBonusRound   `json:"bonus_round,omitempty"`
}

type GameWalletBonusRound struct {
	RoundID         string                  `json:"round_id"`
	Kind            string                  `json:"kind"`
	Token           string                  `json:"token,omitempty"`
	ExpiresAt       time.Time               `json:"expires_at"`
	Choices         int                     `json:"choices"`
	Status          string                  `json:"status,omitempty"`
	BetCredits      string                  `json:"bet_credits,omitempty"`
	BonusCount      int                     `json:"bonus_count,omitempty"`
	BonusMultiplier int                     `json:"bonus_multiplier,omitempty"`
	Options         []GameWalletBonusOption `json:"options,omitempty"`
}

type GameWalletBonusOption struct {
	Choice     int `json:"choice"`
	Spins      int `json:"spins"`
	Multiplier int `json:"multiplier"`
}

type GameWalletBonusClaimRequest struct {
	Token  string `json:"token" binding:"required"`
	Choice *int   `json:"choice" binding:"required"`
}

type GameWalletBonusClaimResult struct {
	RoundID          string                      `json:"round_id"`
	Kind             string                      `json:"kind"`
	Choice           int                         `json:"choice"`
	RewardCredits    string                      `json:"reward_credits"`
	CreditsAfter     string                      `json:"credits_after"`
	ClaimedAt        time.Time                   `json:"claimed_at"`
	IdempotentReplay bool                        `json:"idempotent_replay"`
	Spins            int                         `json:"spins,omitempty"`
	Multiplier       int                         `json:"multiplier,omitempty"`
	BonusMultiplier  int                         `json:"bonus_multiplier,omitempty"`
	TotalBasePayout  string                      `json:"total_base_payout,omitempty"`
	SpinResults      []GameWalletBonusSpinResult `json:"spin_results,omitempty"`
}

type GameWalletBonusSpinResult struct {
	BetCredits       string                  `json:"bet_credits"`
	PayoutCredits    string                  `json:"payout_credits"`
	Grid             [][]string              `json:"grid"`
	Stops            []int                   `json:"stops"`
	WinningLines     []GameWalletWinningLine `json:"winning_lines"`
	ScatterCount     int                     `json:"scatter_count"`
	BonusCount       int                     `json:"bonus_count"`
	FreeSpinsAwarded int                     `json:"free_spins_awarded"`
	PaytableVersion  string                  `json:"paytable_version"`
	ReelStripVersion string                  `json:"reel_strip_version"`
}

type GameWalletScoreRequest struct {
	GameID string            `json:"game_id" binding:"required"`
	Score  *GameWalletNumber `json:"score" binding:"required"`
}

type GameWalletLeaderboardItem struct {
	Rank          int64     `json:"rank"`
	UserDisplay   string    `json:"user_display"`
	Score         string    `json:"score"`
	AchievedAt    time.Time `json:"achieved_at"`
	IsCurrentUser bool      `json:"is_current_user"`
}

type GameWalletLeaderboard struct {
	GameID    string                      `json:"game_id"`
	LocalDate string                      `json:"local_date"`
	Items     []GameWalletLeaderboardItem `json:"items"`
	MyEntry   *GameWalletLeaderboardItem  `json:"my_entry"`
}

type GameWalletRewardClaimRequest struct {
	RewardID string `json:"reward_id"`
}

type GameWalletRewardItem struct {
	ID             string `json:"id"`
	Title          string `json:"title"`
	CreditCost     string `json:"credit_cost"`
	VoucherValue   string `json:"voucher_value"`
	DailyStock     int    `json:"daily_stock"`
	ExpiryDays     int    `json:"expiry_days"`
	RemainingStock *int   `json:"remaining_stock"`
	ClaimedToday   int    `json:"claimed_today"`
}

type GameWalletRewardList struct {
	Items            []GameWalletRewardItem `json:"items"`
	DailyRewardLimit int                    `json:"daily_reward_limit"`
	ClaimedToday     int                    `json:"claimed_today"`
	LoyaltyEnabled   bool                   `json:"loyalty_enabled"`
	LoyaltyAvailable bool                   `json:"loyalty_available"`
}

type GameWalletRewardClaimResult struct {
	ID               string     `json:"id"`
	UserID           int64      `json:"user_id"`
	RewardID         string     `json:"reward_id"`
	Title            string     `json:"title"`
	CreditCost       string     `json:"credit_cost"`
	VoucherValue     string     `json:"voucher_value"`
	RedeemCode       string     `json:"redeem_code"`
	ClaimLocalDate   string     `json:"claim_local_date"`
	CreditsBefore    string     `json:"credits_before"`
	CreditsAfter     string     `json:"credits_after"`
	ExpiresAt        *time.Time `json:"expires_at,omitempty"`
	IdempotencyKey   string     `json:"idempotency_key"`
	IdempotentReplay bool       `json:"idempotent_replay,omitempty"`
	CreatedAt        time.Time  `json:"created_at"`
}

type GameCreditGiftRequest struct {
	RecipientEmail string            `json:"recipient_email" binding:"required"`
	Credits        *GameWalletNumber `json:"credits" binding:"required"`
}

type GameCreditGiftResult struct {
	ID                     string    `json:"id"`
	SenderUserID           int64     `json:"sender_user_id"`
	RecipientUserID        int64     `json:"recipient_user_id"`
	RecipientEmail         string    `json:"recipient_email"`
	Credits                string    `json:"credits"`
	SenderCreditsBefore    string    `json:"sender_credits_before"`
	SenderCreditsAfter     string    `json:"sender_credits_after"`
	RecipientCreditsBefore string    `json:"recipient_credits_before"`
	RecipientCreditsAfter  string    `json:"recipient_credits_after"`
	IdempotencyKey         string    `json:"idempotency_key"`
	IdempotentReplay       bool      `json:"idempotent_replay,omitempty"`
	CreatedAt              time.Time `json:"created_at"`
}

// GameWalletTransaction is now a credit-ledger entry (R21). Legacy exchange
// fields are omitted; clients should use entry_type + amount.
type GameWalletTransaction struct {
	ID             string          `json:"id"`
	UserID         int64           `json:"user_id"`
	EntryType      string          `json:"entry_type"`
	Amount         string          `json:"amount"`
	CreditsBefore  string          `json:"credits_before"`
	CreditsAfter   string          `json:"credits_after"`
	ReferenceType  string          `json:"reference_type,omitempty"`
	ReferenceID    *string         `json:"reference_id,omitempty"`
	IdempotencyKey string          `json:"idempotency_key"`
	Metadata       json.RawMessage `json:"metadata,omitempty"`
	CreatedAt      time.Time       `json:"created_at"`
}

type GameWalletTransactionList struct {
	Items        []GameWalletTransaction `json:"items"`
	NextBeforeID *string                 `json:"next_before_id"`
}
